### PR TITLE
feat(envd): Add multi-part upload API endpoints

### PR DESCRIPTION
**Pull request by mishushakov**
## Summary

Implement multi-part upload in envd (the in-VM daemon) similar to Google Cloud Storage composite upload. This allows large files to be uploaded in parallel parts and then assembled using zero-copy writes.

## Features

- **POST /files/upload/init**: Initialize upload session with UUID
  - Returns unique uploadId for the session
  - Validates user and resolves file path
  
- **PUT /files/upload/{uploadId}?partNumber=n**: Upload individual parts
  - Client-specified part numbers enable parallel uploads
  - Parts are stored in temporary directory with zero-padded naming
  - Supports re-uploading a part (idempotent operation)
  
- **POST /files/upload/{uploadId}/complete**: Assemble and finalize
  - Reads all parts in sorted order
  - Creates parent directories with proper ownership
  - Assembles using zero-copy writes (copy_file_range on Linux via file.ReadFrom)
  - Cleans up temporary directory and session metadata
  
- **DELETE /files/upload/{uploadId}**: Abort and cleanup
  - Removes temporary directory
  - Cleans up session metadata

## Implementation Details

- **Session Storage**: In-memory sync.Map on API struct for concurrent-safe state tracking
- **Temporary Storage**: /tmp/multipart-uploads/{uploadId}/ with zero-padded part files (000000, 000001, etc.)
- **No Session Expiration**: Fully stateless design as per requirements
- **Parallel Upload Support**: Client controls part numbers, enabling safe parallel uploads
- **Zero-Copy Assembly**: Uses Go's file.ReadFrom() which dispatches to copy_file_range syscall on Linux
- **Error Handling**: Proper HTTP status codes (404 for missing uploads, 507 for disk space, etc.)

## Testing

Created comprehensive test suite with 14 test cases covering:
- ✅ Session initialization and cleanup
- ✅ Part upload (single and multiple)
- ✅ Part re-upload (overwrite)
- ✅ Complete operation with assembly
- ✅ Parent directory creation
- ✅ Abort operation
- ✅ Large files (10 × 1MB parts)
- ✅ Empty files
- ✅ Out-of-order part uploads
- ✅ Round-trip testing (upload via multipart, download via GetFiles)
- ✅ Error conditions (missing uploads, insufficient storage, etc.)

## Files Changed

- `packages/envd/spec/envd.yaml`: Added 4 new endpoints and 3 schemas
- `packages/envd/internal/api/api.gen.go`: Auto-generated types and router entries
- `packages/envd/internal/api/multipart.go`: Handler implementations (NEW)
- `packages/envd/internal/api/multipart_test.go`: Comprehensive test suite (NEW)
- `packages/envd/internal/api/store.go`: Added uploads sync.Map field

## Key Design Decisions

1. **In-Memory Metadata**: Uses sync.Map instead of filesystem storage for simplicity and performance
2. **Client Part Numbering**: Enables parallel uploads without server coordination
3. **Zero-Copy Assembly**: Leverages OS-level copy_file_range for efficiency
4. **No UUID Validation**: Relies on OpenAPI format: uuid validation from oapi-codegen
5. **Access Token Auth**: Uses existing auth pattern from fileserving endpoints

🤖 Generated with Claude Code

---

> [!NOTE]
> **Medium Risk**
> Adds new file-write endpoints and concurrency/temporary-file handling, which can impact disk usage, correctness, and cleanup under load. Existing upload/download paths are largely untouched, but the new API surface and filesystem operations increase operational risk.
> 
> **Overview**
> Adds a new multipart upload workflow to `envd` so large files can be uploaded in parallel chunks and then atomically assembled server-side, with endpoints to initialize an upload session, upload numbered parts, complete (assemble + rename) the final file, and abort/cleanup. This introduces in-memory upload session tracking plus temp disk storage under `/tmp`, updates the OpenAPI spec and generated server/client code accordingly, and adds a comprehensive test suite covering correctness and concurrency edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d20534a1acb046a2d10cb3b5aec6e979a39721ab. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/envd/internal/api/multipart.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+	err = os.MkdirAll(uploadDir(uploadId), 0o700)
+	if err != nil {
+		errMsg := fmt.Errorf("error creating upload directory: %w", err)
+		a.logger.Error().Err(errMsg).Str(string(logs.OperationIDKey), operationID).Msg("failed to create upload dir")
+		jsonError(w, http.StatusInternalServerError, errMsg)
+
+		return
+	}
+
+	a.uploads.Store(uploadId, &uploadMetadata{
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/envd/internal/api/multipart.go]: I don't want to do this on purpose, because Sandboxes tend to be shorter lived

File: packages/envd/internal/api/multipart.go [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+	if _, loaded := a.uploads.LoadAndDelete(uploadId); !loaded {
+		meta.mu.Unlock()
+
+		return nil, fmt.Errorf("upload session not found")
+	}
+
+	return meta, nil
+}
+
+func (a *API) PostFilesUploadInit(w http.ResponseWriter, r *http.Request, params PostFilesUploadInitParams) {
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/envd/internal/api/multipart.go]: we can ignore signing here

File: packages/envd/internal/api/multipart.go [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+		errMsg := fmt.Errorf("error resolving path: %w", err)
+		a.logger.Error().Err(errMsg).Str(string(logs.OperationIDKey), operationID).Msg("path resolution failed")
+		jsonError(w, http.StatusBadRequest, errMsg)
+
+		return
+	}
+
+	uploadId := uuid.New().String()
+
+	err = os.MkdirAll(uploadDir(uploadId), 0o700)
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/envd/internal/api/multipart.go]: we never reboot the VM's so that's okay
